README.md

```diff
@@ -24,10 +24,27 @@ def application do
 end
 
 def deps do
-  [{:slack, "~> 0.18.0"}]
+  [{:slack, "~> 1.0.0"}]
 end
 ```
 
+## Upgrading from 0.x to 1.0
+
+The newest version of the Slack client introduces breaking changes with regards to starting and connecting to the Real Time Messaging API. `rtm.start` is now [deprecated](https://api.slack.com/methods/rtm.start) and has since been replaced with [`rtm.connect`](https://api.slack.com/methods/rtm.connect). **This has removed the list of  `bots`, `channels`, `groups`, `users`, and `ims` that are normally returned from `rtm.start`**. Additionally, these lists are now rate-limited. In order to achieve relative parity to the old way of doing things, you'll need to make one change in your code:
+
+### Make additional calls to the Slack API fo grab bots, channels, groups, users, and IMs
+
+Wherever you grab the passed in `slack` state, add in additional calls to populate these lists:
+
+```elixir
+slack
```

Comment:
`websocket_handle` in `Slack.Bot` is calling `Slack.State.update_slack`, which attempts to keep these maps up to date.

This creates a race condition where an update might arrive before the maps are created resulting in:
```
[error] ** Websocket client Slack.Bot terminating in :websocket_handle/3
   for the reason :error:{:badmap, nil}
```

Either the autoupdate routines should be purged or the system needs to be initialized with empty maps in place.

Reply:
oh good point, I would recommend purging them since those maps are no longer needed. Will save connections and also improve performance.

Reply:
Sorry if this was already discussed, but why wouldn't we grab this data for the user? Having to manually manage these maps isn't a great experience imo.

Reply:
Agreed. Existing code can be transformed into a battle tested example documentation.

I'll work on a cherry-pickable commit for this in a few hours.

Reply:
@BlakeWilliams We can do that, but IMO it shouldn't be done in a structure that gets passed around. What do you think about an agent or ETS backed store with common access interface?

Reply:
@acconrad please take a look at Slotos/Elixir-Slack@703bf400dae12effc4c9f1e6f92439a90b9dee6c

It introduces non-nil defaults to `Slack.State` and updates update routines to not fail when they encounter those. Tests pass, although I would (and will try to) spend more time to increase test coverage, as some cases aren't covered properly. There's likely to be a better way to handle some of the updates, but being an Elixir newbie I'm likely to be blind to them.

Next step is to remove updates handling from `Slack.Bot` and transform them into documentation examples for `handle_` callbacks in specific bots.

@BlakeWilliams There is an axolotl-sized dragon lying in `Slack.State` structure. While the data is being fetched an update can arrive with fresher state, which will then be promptly overridden by fetched bulk-update with older data. Slack records and events carry timestamps to indicate freshness of the data, but it's getting lost. In larger organizations this is a likely race condition with potentially visible, if not too significant, outcomes. Still, I'd personally leave implementation of a properly synchronized storage to whoever writes the bot, at least for the 1.0.0 release.

Reply:
> Still, I'd personally leave implementation of a properly synchronized storage to whoever writes the bot, at least for the 1.0.0 release.

I feel like maintaining the state is a fundamental part of the library. It lets the folks writing bots worry about real issues instead of the plumbing to keep things up-to-date. It also feels like something that will be re-invented many times over if not implemented.

I wouldn't be opposed to it being optional, if anyone has any compelling use cases.

Reply:
> @BlakeWilliams There is an axolotl-sized dragon lying in Slack.State structure. While the data is being fetched an update can arrive with fresher state, which will then be promptly overridden by fetched bulk-update with older data.

I think that's a good point. It doesn't seem like a *huge* issue, but I definitely think it should be solved before we `1.0` the library. One thought would be that before we connect, we fetch the data to bootstrap the state before connecting.